### PR TITLE
Refactor version checks

### PR DESCRIPTION
**Pull request by antstorm**
This PR adds 2 little function for version checks leveraging already included `semver` and getting rid of duplication.

The only bit I'm not sure about is

``` javascript
if (CONFIG.flightControllerIdentifier === "BTFL" || semver.ge(CONFIG.flightControllerVersion, "2.9.0")) {
  $('#pid_filter').show();
}
```

I guess `||` here is a typo, right?

### CODE REVIEW

**Comment by antstorm** [comment]
@mikeller would be good to get your eyes on this one, cheers!

File: js/serial_backend.js [hunk 1]
```diff
@@ -184,7 +184,7 @@ function onOpen(openInfo) {
         MSP.send_message(MSP_codes.MSP_API_VERSION, false, false, function () {
             GUI.log(chrome.i18n.getMessage('apiVersionReceived', [CONFIG.apiVersion]));
 
-            if (semver.gte(CONFIG.apiVersion, CONFIGURATOR.apiVersionAccepted)) {
+            if (FC.apiVersion('>=' + CONFIGURATOR.apiVersionAccepted)) {
```
Comment (mikeller) [inline js/serial_backend.js]: This is unsafe. If `satisfies` is used, it should be with constants.

File: tabs/setup.js [hunk 2]
```diff
@@ -37,7 +37,7 @@ TABS.setup.initialize = function (callback) {
         // translate to user-selected language
         localize();
 
-        if (semver.lt(CONFIG.apiVersion, CONFIGURATOR.backupRestoreMinApiVersionAccepted)) {
+        if (FC.apiVersion('<' + CONFIGURATOR.backupRestoreMinApiVersionAccepted)) {
```
Comment (mikeller) [inline tabs/setup.js]: This is unsafe. If `satisfies` is used, it should be with constants.

**Comment by mikeller** [comment]
> I guess || here is a typo, right?

It looks like it.

I like the idea of providing wrapper functions for version checks that makes them a bit neater to write. What I do not like is the move form using expclicit comparison operators `.eq`, `.gt`, `.gte` to using string evaluation for comparison. This is a regression, it goes from more precise to less precise, and it also makes the code less accessible to manipulation by IDE.

My suggestion is to normalise the checks to only use one set of comparison operations (i.e. `.eq`, `.gt`,  `.gte`), and add these to the wrapper that is used for comparison, in place of the comparison by `.satisfies`.

Another thing is that I think it makes sense to stop supporting non-Betaflight altogether, and not even allow non - Betaflight FCs to connect to the configurator. In this case, we can remove all of the `CONFIG.flightControllerIdentifier == "BTFL"` checks. @borisbstyle might want to weigh in on the question of removing support for non - Betaflight FCs though.

Also, we should add support for checking the api version to the version check wrapper:

https://github.com/antstorm/betaflight-configurator/blob/f7417f4e39834fe00a2ec93e7bcab3fd6c23535b/js/backup_restore.js#L26

**Comment by antstorm** [comment]
> I like the idea of providing wrapper functions for version checks that makes them a bit neater to write. What I do not like is the move form using expclicit comparison operators .eq, .gt, .gte to using string evaluation for comparison. This is a regression, it goes from more precise to less precise, and it also makes the code less accessible to manipulation by IDE.

I'm not sure why you think it's less precise. I think it gives a git more flexibility like `FC.version('2.x')`. There's no direct string evaluation if that's what you're referring to. And it's one of the suggested way in their README. But this can definitely be changed to comparison methods directly.

> Another thing is that I think it makes sense to stop supporting non-Betaflight altogether, and not even allow non - Betaflight FCs to connect to the configurator. In this case, we can remove all of the CONFIG.flightControllerIdentifier == "BTFL" checks. @borisbstyle might want to weigh in on the question of removing support for non - Betaflight FCs though.

Now that's it's contained within the version checker, it's very easy to drop. Do you want me to do it as part of this PR?

> Also, we should add support for checking the api version to the version check wrapper:
> https://github.com/antstorm/betaflight-configurator/blob/f7417f4e39834fe00a2ec93e7bcab3fd6c23535b/js/backup_restore.js#L26

Well, there's `FC.apiVersion('< 1.12.0')` on that line, which users api version check wrapper. Not sure I got your comment right.

**Comment by antstorm** [comment]
Another option is `FC.version('>=', '2.8.2')` with a simple string comparison and direct method call and if only 1 param is given — send it to `satisfies`, allowing for `2.x` stuff.

File: tabs/adjustments.js [hunk 3]
```diff
@@ -68,7 +68,7 @@ TABS.adjustments.initialize = function (callback) {
             availableFunctionCount += 2; // pitch and roll rate
             if (semver.gte(CONFIG.flightControllerVersion, '1.9.0')) {
```
Comment (mikeller) [inline tabs/adjustments.js]: Looks like you missed that one.

**Comment by mikeller** [comment]
> I'm not sure why you think it's less precise.

Probably bad choice of words on my behalf. It's more that expressivity in code is moved into expressivity in strings. 

> Now that's it's contained within the version checker, it's very easy to drop. Do you want me to do it as part of this PR?

No, wait with that until we've got feedback from @borisbstyle. I'll hold off mergeing until he gets back to us.

> Well, there's FC.apiVersion('< 1.12.0') on that line, which users api version check wrapper. Not sure I got your comment right.

Sorry, that was me not being awake yet. Please disregard this comment.

> Another option is FC.version('>=', '2.8.2') with a simple string comparison and direct method call and if only 1 param is given — send it to satisfies, allowing for 2.x stuff.

Still the same really, unless you only allow `=`, `>`, and `>=`, and throw exceptions for all other values as first argument. I'd really like to get people to normalise checks like this, and making it the only option in code beats nagging for it in pull requests.

**Comment by antstorm** [comment]
> Still the same really, unless you only allow =, >, and >=, and throw exceptions for all other values as first argument. I'd really like to get people to normalise checks like this, and making it the only option in code beats nagging for it in pull requests.

Yeah, allowing only `>`, `>=`, `<`, `<=`. `=` is not needed, because you can simply `FC.version('2.8.1')`. Throw exception otherwise and maybe even check that version comparing to consists of nothing but numbers, dots and `x`s.

**Comment by mikeller** [comment]
I didn't quite get your last comment. What I want to achieve is that all version checks are normalised, which makes simplifying them a whole lot easier. So instead of allowing `FC.version('>= 2.8.0')` or `!FC.version('< 2.8.0')`, just allow one of the 2. With your suggestion, this requires extra checks. With my suggestion, all there is to it is to only add `.gte`, `.gt` and `.eq`, but not `.lte` or `.lt`.

**Comment by antstorm** [comment]
Oh, right I see. I used `!FC.version('>= 2.8.0')` as a replacement for `CONFIG.flightControllerIdentifier !== "BTFL" || semver.lt(CONFIG.flightControllerVersion, "2.8.0")`. I was a bit confused by the format of this check… I mean it passes if FC is not BetaFlight OR it's a BetaFlight with version < 2.8.0. I guess this can be simply reduced to `!FC.version('>= 2.8.0')`.

I like the ability to compare with `2.8.x`, so unless you are strongly against it I will restrict comparisons to `>`, `>=` and single param for `eq`.

**Comment by mikeller** [comment]
I am not a huge fan of `2.8.x` comparisons, as they encourage behaviour that will break over time. (i.e. adding a `FC.version('2.8.x')` before version 2.9 has been released, and then later on a fix for 2.8 breaks whatever you are gating for). If you want something that limits releases to lower and upper bounds, you should be explicit about it. What about `FC.apiVersion.between('2.8.0', '2.9.0)` with inclusive, exclusive semantics?

**Comment by borisbstyle** [comment]
The reason I wanted to check for this > CONFIG.flightControllerIdentifier == "BTFL" was because once Cleanflight ever reaches the same version it will also be accepted! Cleanflight is now version 1.13.0 I believe and will catch up once in versioning. Checking for flightcontroller identifier will prevent that.

|| = OR statement and it was not a typo

**Comment by antstorm** [comment]
@borisbstyle 

```javascript
if (CONFIG.flightControllerIdentifier === "BTFL" || semver.ge(CONFIG.flightControllerVersion, "2.9.0")) {
  $('#pid_filter').show();
}
```

But you can not rely on Cleanflight having the same features in the future version `2.9.0`, right? I think a bigger question is — can we just have one check when connecting to FC and after that assume it's BetaFlight?

**Comment by antstorm** [comment]
@mikeller I've addressed your comments, but now it doesn't rely on `CONFIG. flightControllerIdentifier ` check. Is there some agreement on what should we do with that check?

**Comment by borisbstyle** [comment]
@antstorm 
We dont want the users with cleanflight who want to connect the configurator to break stuff. Maybe only one check during connect would be fine? If no BTFL not allowed to connect?

**Comment by antstorm** [comment]
@borisbstyle that's exactly what I thought. I'll add that check. Also @mikeller mentioned that there's no good way of identifying early version of BetaFlight. Do we need to worry about it?

**Comment by mikeller** [comment]
@borisbstyle that's my thoughts too. Looking back through releases, the first release for Betaflight seems to have been 2.1.0. If we allow connections only from FCs running Betaflight, we can clean up by removing any checks and code for older versions than 2.1.0.

**Comment by hydra** [comment]
see also this commit:

https://github.com/cleanflight/cleanflight-configurator/commit/305f827654057618858e10f48a0473fd50d748ca

**Comment by mikeller** [comment]
That's great @hydra. Do you think this can be made into a PR for the Betaflight configurator?

**Comment by hydra** [comment]
@mikeller I've not currently got time, just merge the commits from this PR merge.

https://github.com/cleanflight/cleanflight-configurator/commit/ff7478df68d9050977ca2210459269dd277cd521

**Comment by hydra** [comment]
@mikeller i think you can close this now.

**Pull request closed without merging**